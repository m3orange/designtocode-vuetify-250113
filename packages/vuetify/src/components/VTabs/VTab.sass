@use './variables' as *

.v-tab
  // override v-btn size specificity
  &.v-tab
    --v-btn-height: var(--v-tabs-height)
    min-width: $tab-min-width

  .v-tabs--horizontal &
    max-width: $tab-max-width

  .v-tabs--vertical &
    justify-content: start

.v-tab__slider
  position: absolute
  bottom: 0
  left: 0
  height: $tab-slider-size
  width: 100%
  background: currentColor
  pointer-events: none
  opacity: 0

  .v-tab--selected &
    opacity: 1

  .v-tabs--vertical &
    top: 0
    height: 100%
    width: $tab-slider-size
