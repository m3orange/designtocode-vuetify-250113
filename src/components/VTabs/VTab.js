// Mixins
import Routable from '../../mixins/routable'
import {
  inject as RegistrableInject
} from '../../mixins/registrable'

export default {
  name: 'v-tab',

  mixins: [
    RegistrableInject('tabs', 'v-tab', 'v-tabs'),
    Routable
  ],

  inject: ['tabClick'],

  data () {
    return {
      isActive: false
    }
  },

  props: {
    activeClass: {
      type: String,
      default: 'tabs__item--active'
    },
    ripple: {
      type: [Boolean, Object],
      default: true
    }
  },

  computed: {
    classes () {
      return {
        'tabs__item': true,
        'tabs__item--disabled': this.disabled,
        [this.activeClass]: !this.to && this.isActive
      }
    },
    action () {
      const to = this.to || this.href

      if (!to || to === Object(to)) return this._uid

      return to.replace('#', '')
    }
  },

  watch: {
    $route: 'onRouteChange'
  },

  mounted () {
    this.tabs.register({
      id: this.action,
      toggle: this.toggle,
      el: this.$el
    })
  },

  beforeDestroy () {
    this.tabs.unregister(this.action)
  },

  methods: {
    click (e) {
      // If user provides an
      // actual link, do not
      // prevent default
      if (this.href &&
        this.href.indexOf('#') > -1
      ) e.preventDefault()

      this.$emit('click', e)

      if (!this.to && !this.href) return
      if (!this.to) this.tabClick(this.action)
    },
    onRouteChange () {
      if (!this.to) return

      this.$nextTick(() => {
        if (this.$el.firstChild.className.indexOf(this.activeClass) > -1) {
          this.tabClick(this.action)
        }
      })
    },
    toggle (action) {
      this.isActive = this.action === action
    }
  },

  render (h) {
    const link = this.generateRouteLink()
    const { data } = link

    // If disabled, use div as anchor tags do not support
    // being disabled
    const tag = this.disabled ? 'div' : link.tag

    return h('div', {
      staticClass: 'tabs__div'
    }, [h(tag, data, this.$slots.default)])
  }
}
