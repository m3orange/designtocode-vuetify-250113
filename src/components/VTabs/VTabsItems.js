// Mixins
import {
  provide as RegistrableProvide
} from '../../mixins/registrable'

// Directives
import Touch from '../../directives/touch'

export default {
  name: 'v-tabs-items',

  mixins: [RegistrableProvide('tabs')],

  directives: { Touch },

  inject: {
    registerItems: {
      default: null
    },
    unregisterItems: {
      default: null
    }
  },

  data () {
    return {
      items: [],
      lazyValue: this.value,
      reverse: false
    }
  },

  props: {
    cycle: Boolean,
    touchless: Boolean,
    value: null
  },

  computed: {
    activeIndex () {
      return this.items.findIndex(item => item.id === this.lazyValue)
    },
    activeItem () {
      if (!this.items.length) return undefined

      return this.items[this.activeIndex]
    },
    inputValue: {
      get () {
        return this.lazyValue
      },
      set (val) {
        this.lazyValue = val

        !this.registerItems && this.$emit('input', val)
      }
    }
  },

  watch: {
    activeIndex (current, previous) {
      this.reverse = current < previous
      this.updateItems()
    },
    value (val) {
      this.lazyValue = val
    }
  },

  mounted () {
    this.registerItems && this.registerItems(this.changeModel)
  },

  beforeDestroy () {
    this.unregisterItems && this.unregisterItems()
  },

  methods: {
    changeModel (val) {
      this.inputValue = val
    },
    next (cycle) {
      let nextIndex = this.activeIndex + 1

      if (!this.items[nextIndex]) {
        if (!cycle) return
        nextIndex = 0
      }

      this.inputValue = this.items[nextIndex].id
    },
    prev (cycle) {
      let prevIndex = this.activeIndex - 1

      if (!this.items[prevIndex]) {
        if (!cycle) return
        prevIndex = this.items.length - 1
      }

      this.inputValue = this.items[prevIndex].id
    },
    onSwipe (action) {
      this[action](this.cycle)
    },
    register (options) {
      this.items.push(options)
    },
    unregister (id) {
      this.items = this.items.filter(item => item.id !== id)
    },
    updateItems () {
      this.items.forEach(({ toggle }) => {
        toggle(this.lazyValue, this.reverse, this.isBooted)
      })
      this.isBooted = true
    }
  },

  render (h) {
    const data = {
      staticClass: 'tabs__items',
      directives: []
    }

    !this.touchless && data.directives.push({
      name: 'touch',
      value: {
        left: () => this.onSwipe('next'),
        right: () => this.onSwipe('prev')
      }
    })

    return h('div', data, this.$slots.default)
  }
}
