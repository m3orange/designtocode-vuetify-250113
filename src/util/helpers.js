export function createSimpleFunctional (c, el = 'div', name) {
  name = name || c.replace(/__/g, '-')

  return {
    name: `v-${name}`,
    functional: true,

    render: (h, { data, children }) => {
      data.staticClass = (`${c} ${data.staticClass || ''}`).trim()

      return h(el, data, children)
    }
  }
}

export function createSimpleTransition (name, origin = 'top center 0', mode) {
  return {
    name,

    functional: true,

    props: {
      origin: {
        type: String,
        default: origin
      }
    },

    render (h, context) {
      context.data = context.data || {}
      context.data.props = { name }
      context.data.on = context.data.on || {}
      if (!Object.isExtensible(context.data.on)) {
        context.data.on = { ...context.data.on }
      }

      if (mode) context.data.props.mode = mode

      context.data.on.beforeEnter = el => {
        el.style.transformOrigin = context.props.origin
        el.style.webkitTransformOrigin = context.props.origin
      }

      return h('transition', context.data, context.children)
    }
  }
}

export function createJavaScriptTransition (name, functions, css = true, mode = 'in-out') {
  return {
    name,

    functional: true,

    props: {
      css: {
        type: Boolean,
        default: css
      },
      mode: {
        type: String,
        default: mode
      }
    },

    render (h, context) {
      const data = {
        props: {
          ...context.props,
          name
        },
        on: functions
      }

      return h('transition', data, context.children)
    }
  }
}

export function directiveConfig (binding, defaults = {}) {
  return Object.assign({},
    defaults,
    binding.modifiers,
    { value: binding.arg },
    binding.value || {}
  )
}

export function addOnceEventListener (el, event, cb) {
  var once = () => {
    cb()
    el.removeEventListener(event, once, false)
  }

  el.addEventListener(event, once, false)
}

export function getNestedValue (obj, path) {
  for (let i = 0, m = path.length; i < m; i++) {
    if (obj === null || obj === undefined) {
      return
    }
    obj = obj[path[i]]
  }
  return obj;
}

export function getObjectValueByPath (obj, path) {
  // credit: http://stackoverflow.com/questions/6491463/accessing-nested-javascript-objects-with-string-key#comment55278413_6491621
  if (!path || path.constructor !== String) return
  path = path.replace(/\[(\w+)\]/g, '.$1') // convert indexes to properties
  path = path.replace(/^\./, '') // strip a leading dot
  return getNestedValue(obj, path.split('.'))
}

export function createRange (length) {
  return [...Array.from({ length }, (v, k) => k)]
}

export function getZIndex (el) {
  if (!el || el.nodeType !== Node.ELEMENT_NODE) return 0
  var zi = document.defaultView.getComputedStyle(el).getPropertyValue('z-index')
  if (isNaN(zi)) return getZIndex(el.parentNode)

  return zi
}

const tagsToReplace = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;'
}

export function escapeHTML (str) {
  return str.replace(/[&<>]/g, tag => tagsToReplace[tag] || tag)
}

export function looseEqual(a, b) {
    if (a === b) return true
    if (typeof a !== 'object' || typeof b !== 'object') {
        // If the values aren't objects, they were already checked for equality
        return false
    }
    const props = Object.keys(a)
    if (props.length !== Object.keys(b).length) {
        // Different number of props, don't bother to check
        return false
    }
    return props.every(p => looseEqual(a[p], b[p]));
}

export function filterObjectOnKeys(obj, keys) {
  const filtered = {}

  for (let i = 0; i < keys.length; i++) {
    const key = keys[i]
    if (typeof obj[key] !== 'undefined') {
      filtered[key] = obj[key]
    }
  }

  return filtered
}

export function filterChildren (array = [], tag) {
  return array.filter(child => {
    return child.componentOptions &&
      child.componentOptions.Ctor.options.name === tag
  })
}
