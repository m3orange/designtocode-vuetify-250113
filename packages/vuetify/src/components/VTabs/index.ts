import VTabs from './VTabs'
import VTab from './VTab'
import VTabsItems from './VTabsItems'
import VTabItem from './VTabItem'
import VTabsSlider from './VTabsSlider'
import VTabsBar from './VTabsBar'

export { VTabs, VTab, VTabItem, VTabsItems, VTabsSlider, VTabsBar }

export default {
  $_vuetify_subcomponents: {
    VTabs,
    VTab,
    VTabsItems,
    VTabItem,
    VTabsSlider,
    VTabsBar,
  },
}
