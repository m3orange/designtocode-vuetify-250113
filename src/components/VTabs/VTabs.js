// Styles
require('../../stylus/components/_tabs.styl')

// Component imports
import VIcon from '../VIcon'
import VTabsItems from './VTabsItems'
import VTabsSlider from './VTabsSlider'

// Component level mixins
import TabsComputed from './mixins/tabs-computed'
import TabsGenerators from './mixins/tabs-generators'
import TabsProps from './mixins/tabs-props'
import TabsTouch from './mixins/tabs-touch'
import TabsWatchers from './mixins/tabs-watchers'

// Mixins
import Colorable from '../../mixins/colorable'
import SSRBootable from '../../mixins/ssr-bootable'
import Themeable from '../../mixins/themeable'
import {
  provide as RegistrableProvide
} from '../../mixins/registrable'

// Directives
import Resize from '../../directives/resize'
import Touch from '../../directives/touch'

export default {
  name: 'v-tabs',

  components: {
    VIcon,
    VTabsItems,
    VTabsSlider
  },

  mixins: [
    RegistrableProvide('tabs'),
    Colorable,
    SSRBootable,
    TabsComputed,
    TabsProps,
    TabsGenerators,
    TabsTouch,
    TabsWatchers,
    Themeable
  ],

  directives: {
    Resize,
    Touch
  },

  provide () {
    return {
      tabClick: this.tabClick,
      registerItems: this.registerItems,
      unregisterItems: this.unregisterItems
    }
  },

  data () {
    return {
      prependIconVisible: false,
      appendIconVisible: false,
      bar: [],
      content: [],
      isBooted: false,
      isOverflowing: false,
      lazyValue: this.value,
      resizeTimeout: null,
      reverse: false,
      scrollOffset: 0,
      sliderWidth: null,
      sliderLeft: null,
      startX: 0,
      tabsContainer: null,
      tabs: [],
      tabItems: null,
      transitionTime: 300
    }
  },

  methods: {
    checkPrependIcon () {
      return this.scrollOffset > 0
    },
    checkAppendIcon () {
      // Check one scroll ahead to know the width of right-most item
      const container = this.$refs.container
      const wrapper = this.$refs.wrapper

      return container.clientWidth > this.scrollOffset + wrapper.clientWidth
    },
    callSlider () {
      this.setOverflow()
      if (!this.activeTab) return false

      // Give screen time to paint
      this.$nextTick(() => {
        const activeTab = this.$el.querySelector(`[href="#${this.activeTab.id}"]`)

        if (!activeTab) return
        this.sliderWidth = activeTab.scrollWidth
        this.sliderLeft = activeTab.offsetLeft
      })
    },
    /**
     * When v-navigation-drawer changes the
     * width of the container, call resize
     * after the transition is complete
     */
    onContainerResize () {
      clearTimeout(this.resizeTimeout)
      this.resizeTimeout = setTimeout(this.callSlider, this.transitionTime)
    },
    onResize () {
      if (this._isDestroyed) return

      this.callSlider()

      if (!this.activeTab) return

      this.scrollOffset = this.activeTab.el.offsetLeft
    },
    overflowCheck (e, fn) {
      this.isOverflowing && fn(e)
    },
    scrollTo (direction) {
      this.scrollOffset = this.newOffset(direction)
    },
    setOverflow () {
      this.isOverflowing = this.$refs.bar.clientWidth < this.$refs.container.clientWidth
    },
    findActiveLink () {
      if (!this.tabs.length || this.lazyValue) return

      const activeIndex = this.tabs.findIndex(tabItem => {
        return tabItem.id === this.lazyValue ||
          tabItem.el.firstChild.className.indexOf(this.activeClass) > -1
      })

      /* istanbul ignore next */
      // There is not a reliable way to test
      this.inputValue = this.tabs[activeIndex > -1 ? activeIndex : 0].id
    },
    parseNodes () {
      const item = []
      const items = []
      const slider = []
      const tab = []
      const length = (this.$slots.default || []).length

      for (let i = 0; i < length; i++) {
        const vnode = this.$slots.default[i]

        /* istanbul ignore else */
        if (vnode.componentOptions) {
          switch (vnode.componentOptions.Ctor.options.name) {
            case 'v-tab': tab.push(vnode)
              break
            case 'v-tabs-slider': slider.push(vnode)
              break
            case 'v-tabs-items': items.push(vnode)
              break
            case 'v-tab-item': item.push(vnode)
              break
          }
        }
      }

      return { tab, slider, items, item }
    },
    register (options) {
      this.tabs.push(options)
    },
    tabClick (target) {
      this.inputValue = target
    },
    registerItems (fn) {
      this.tabItems = fn
    },
    unregisterItems () {
      this.tabItems = null
    },
    unregister (id) {
      this.tabs = this.tabs.filter(o => o.id !== id)
    },
    updateTabs () {
      this.tabs.forEach(({ toggle }) => {
        toggle(this.target)
      })

      this.setOverflow()
    }
  },

  mounted () {
    this.callSlider()
    this.prependIconVisible = this.checkPrependIcon()
    this.appendIconVisible = this.checkAppendIcon()
  },

  render (h) {
    const { tab, slider, items, item } = this.parseNodes()

    return h('div', {
      staticClass: 'tabs',
      directives: [{
        name: 'resize',
        arg: 400,
        modifiers: { quiet: true },
        value: this.onResize
      }]
    }, [
      this.genBar([this.genSlider(slider), tab]),
      this.genItems(items, item)
    ])
  }
}
