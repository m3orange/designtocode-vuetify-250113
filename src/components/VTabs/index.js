import VTabs from './VTabs'
import VTab from './VTab'
import VTabsItems from './VTabsItems'
import VTabItem from './VTabItem'
import VTabsSlider from './VTabsSlider'

export { VTabs, VTabItem, VTab, VTabsItems, VTabsSlider }

VTabs.$_vuetify_subcomponents = {
  VTabsItems,
  VTabItem,
  VTabsSlider
}

export default VTabs
