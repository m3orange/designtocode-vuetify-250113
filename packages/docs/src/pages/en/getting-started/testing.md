---
nav: Testing
meta:
  title: Testing
  description: Details for v3 beta release - faq, changes, and upgrading.
  keywords: test, testing, vitest, jest, vue 3, vuetify 3
---

<script setup>
  import { ref } from 'vue'

  const tab = ref('testing-library')
</script>

# Testing

Introduction here

## End-to-end testing

### Cypress

TBW

## Unit testing

We recommend [vitest](https://vitest.dev/) for your unit testing needs, if you are running a vite based project.

### Vitest

Start by installing `vitest` and the `jsdom` environment.

```bash
yarn -D vitest jsdom
```

Then configure vitest using the `test` object in the vite config. Make sure to add `vuetify` as an inline dependency.

```ts { data-resource="vite.config.ts" }
...

export default defineConfig({
  ...,
  test: {
    globals: true,
    environment: 'jsdom',
    deps: {
      inline: ['vuetify']
    }
  }
})
```

That should be enough to run tests. Go to the [Writing tests](#writing-tests) section to see how to render and test Vuetify components

### Jest

If you already have an existing jest setup that you want to continue using, have a look at [vite-jest](https://github.com/sodatea/vite-jest), which offers first-class Vite integration for jest.

### Writing tests

Here is a simple example component with a **VBtn** component.

```html { data-resource="Hello.vue" }
<script setup lang="ts">
    import { computed, ref } from 'vue'
    import { VBtn } from 'vuetify/components'

    const props = defineProps<{ count: number }>()
    const times = ref(2)
    const result = computed(() => props.count * times.value)

    defineExpose(props)
</script>

<template>
    <div>{{ count }} x {{ times }} = {{ result }}</div>
    <VBtn @click="times += 1">
    x1
    </VBtn>
</template>
```

To render Vue components in unit tests, you can use either [@testing-library/vue](https://github.com/testing-library/vue-testing-library) or [@vue/test-utils](https://github.com/vuejs/test-utils). You can find a good comparison of the two in the [Vue documentation](https://vuejs.org/guide/scaling-up/testing.html#mounting-libraries).

There is not much difference between the two when it comes to mounting Vuetify components. In the examples below we have created a helper function that registers Vuetify as a global plugin.

<v-tabs v-model="tab" color="primary">
  <v-tab value="testing-library" variant="plain">@testing-library/vue</v-tab>
  <v-tab value="test-utils" variant="plain">@vue/test-utils</v-tab>
</v-tabs>

<v-window v-model="tab">
  <v-window-item value="testing-library">

```ts { data-resource="Hello.spec.ts" }
import { render, fireEvent } from '@testing-library/vue'
import { test, expect } from 'vitest'
import { createVuetify } from 'vuetify'
import Hello from './Hello.vue'

function renderWithVuetify (component, options) {
    const vuetify = createVuetify()
    return render(component, {
        ...options,
        global: {
            plugins: [vuetify]
        },
    })
}

test('mount component', async () => {
  expect(Hello).toBeTruthy()

  const wrapper = renderWithVuetify(Hello, {
    props: {
      count: 4,
    },
  })

  expect(wrapper.queryByText('4 x 2 = 8')).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()

  const button = wrapper.queryByText('x1')!

  await fireEvent.click(button)

  expect(wrapper.queryByText('4 x 3 = 12')).toBeTruthy()

  await fireEvent.click(button)

  expect(wrapper.queryByText('4 x 4 = 16')).toBeTruthy()
})
```

  </v-window-item>
  <v-window-item value="test-utils">

```ts { data-resource="Hello.spec.ts" }
import { mount } from '@vue/test-utils'
import { test, expect } from 'vitest'
import { createVuetify } from 'vuetify'
import Hello from './Hello.vue'

function mountWithVuetify (component, options) {
    const vuetify = createVuetify()
    return mount(component, {
        ...options,
        global: {
            plugins: [vuetify]
        },
    })
}

test('mount component', async () => {
  expect(Hello).toBeTruthy()

  const wrapper = mountWithVuetify(Hello, {
    props: {
      count: 4,
    },
  })

  expect(wrapper.text()).toContain('4 x 2 = 8')
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.get('.v-btn').trigger('click')

  expect(wrapper.text()).toContain('4 x 3 = 12')

  await wrapper.get('.v-btn').trigger('click')

  expect(wrapper.text()).toContain('4 x 4 = 16')
})
```

  </v-window-item>
</v-window>





