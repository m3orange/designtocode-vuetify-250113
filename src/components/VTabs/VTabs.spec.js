import { test, touch } from '@util/testing'
import { createRange } from '@util/helpers'
import VTabs from './VTabs'
import VTab from './VTab'
import VTabItem from './VTabItem'
import VTabsItems from './VTabsItems'
import VTabsSlider from './VTabsSlider'

const Component = (items = ['foo', 'bar']) => {
  return {
    inheritAttrs: false,

    render (h) {
      return h(VTabs, {
        attrs: this.$attrs
      }, [
        items.map(item => h(VTab, {
          props: { href: `#${item}` }
        })),
        h(VTabsItems, items.map(item => h(VTabItem, {
          props: { id: item }
        })))
      ])
    }
  }
}

const ssrBootable = () => new Promise(resolve => setTimeout(resolve, 200))

test('VTabs', ({ mount, shallow }) => {
  it('should provide', () => {
    const wrapper = mount(Component())

    const tab = wrapper.find(VTab)[0]
    expect(typeof tab.vm.tabClick).toBe('function')
    expect(typeof tab.vm.tabs.register).toBe('function')
    expect(typeof tab.vm.tabs.unregister).toBe('function')

    const items = wrapper.find(VTabsItems)[0]
    expect(typeof items.vm.registerItems).toBe('function')
    expect(typeof items.vm.unregisterItems).toBe('function')
  })

  it('should register tabs and items', async () => {
    const wrapper = mount(VTabs, {
      slots: {
        default: [VTab, VTabsItems]
      }
    })

    const tab = wrapper.find(VTab)[0]
    expect(wrapper.vm.tabs.length).toBe(1)
    tab.destroy()
    expect(wrapper.vm.tabs.length).toBe(0)

    const items = wrapper.find(VTabsItems)[0]
    expect(typeof wrapper.vm.tabItems).toBe('function')
    items.destroy()
    expect(wrapper.vm.tabItems).toBe(null)
  })

  it('should change tab and content when model changes', async () => {
    const wrapper = mount(Component(), {
      attachToDocument: true
    })

    const tabs = wrapper.find(VTabs)[0]
    const tab = wrapper.find(VTab)[0]
    const item = wrapper.find(VTabItem)[0]

    expect(tabs.vm.activeIndex).toBe(-1)
    expect(tab.vm.isActive).toBe(false)
    expect(item.vm.isActive).toBe(false)
    await ssrBootable()
    expect(tabs.vm.activeIndex).toBe(0)
    expect(tab.vm.isActive).toBe(true)
    expect(item.vm.isActive).toBe(true)
  })

  it('should not call slider if no active tab', () => {
    const wrapper = mount(VTabs)

    expect(wrapper.vm.callSlider()).toBe(false)
  })

  it('should call slider on application resize', async () => {
    const wrapper = mount(Component())

    const tabs = wrapper.find(VTabs)[0]

    expect(tabs.vm.resizeTimeout).toBe(null)
    tabs.vm.$vuetify.application.left = 100
    await tabs.vm.$nextTick()
    expect(tabs.vm.resizeTimeout).toBeTruthy()
    tabs.setData({ resizeTimeout: null })
    expect(tabs.vm.resizeTimeout).toBe(null)
    tabs.vm.$vuetify.application.right = 100
    await tabs.vm.$nextTick()
    expect(tabs.vm.resizeTimeout).toBeTruthy()
  })

  it('should reset offset on resize', async () => {
    const wrapper = mount(Component(), {
      attachToDocument: true
    })

    const tabs = wrapper.find(VTabs)[0]

    tabs.setData({ scrollOffset: 1 })
    tabs.vm.onResize()
    await tabs.vm.$nextTick()
    expect(tabs.vm.scrollOffset).toBe(0)
    tabs.setData({ scrollOffset: 2 })
    await tabs.vm.$nextTick()
    tabs.destroy()
    tabs.vm.onResize()
    expect(tabs.vm.scrollOffset).toBe(2)
  })

  it('should update model when route changes', async () => {
    const $route = { path: 'bar' }
    const wrapper = mount(Component(), {
      attachToDocument: true,
      globals: {
        $route
      }
    })

    await ssrBootable()

    const tabs = wrapper.find(VTabs)[0]
    const tab = wrapper.find(VTab)[1]
    const input = jest.fn()

    tabs.vm.$on('input', input)
    tab.vm.click(new Event('click'))
    await wrapper.vm.$nextTick()

    expect(input).toHaveBeenCalled()
  })

  it('should call method if overflowing', () => {
    const wrapper = mount(VTabs)
    const fn = jest.fn()

    wrapper.vm.overflowCheck(null, fn)
    expect(fn).not.toHaveBeenCalled()
    wrapper.setData({ isOverflowing: true })
    wrapper.vm.overflowCheck(null, fn)
    expect(fn).toHaveBeenCalled()
  })

  it('should update scroll and item offset', async () => {
    const newOffset = jest.fn()
    const wrapper = mount(VTabs, {
      props: {
        showArrows: true
      }
    })

    wrapper.setMethods({ newOffset })

    await ssrBootable()

    wrapper.vm.scrollTo('append')
    wrapper.vm.scrollTo('prepend')
    expect(newOffset.mock.calls.length).toBe(2)

    wrapper.setMethods({ newOffset: () => ({
      offset: 5,
      index: 1
    })})
    await wrapper.vm.$nextTick()

    wrapper.vm.scrollTo('prepend')
    expect(wrapper.vm.scrollOffset).toBe(5)
    expect(wrapper.vm.itemOffset).toBe(1)
  })

  it('should return the correct height', async () => {
    const wrapper = mount(VTabs, {
      propsData: { height: 'auto' }
    })

    expect('Invalid prop: custom validator check failed for prop "height"').toHaveBeenWarned()
    wrapper.setProps({ height: null })
    expect(wrapper.vm.computedHeight).toBe(48)
    wrapper.setProps({ height: 112 })
    expect(wrapper.vm.computedHeight).toBe(112)
    wrapper.setProps({ height: null, iconsAndText: true })
    expect(wrapper.vm.computedHeight).toBe(72)
  })

  it('should return lazy value when accessing input', async () => {
    const wrapper = mount(VTabs)

    expect(wrapper.vm.inputValue).toBe(undefined)
    wrapper.setData({ lazyValue: 'foo' })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.inputValue).toBe('foo')
  })

  it('should show tabs arrows', async () => {
    const wrapper = mount(VTabs, {
      propsData: { showArrows: true }
    })

    wrapper.setData({ isOverflowing: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.hasClass('tabs--show-arrows')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a null target with no activeTab', () => {
    const wrapper = mount(VTabs)

    expect(wrapper.vm.target).toBe(null)
  })

  it('should not conditionally render append and prepend icons', async () => {
    const scrollTo = jest.fn()
    const wrapper = mount(VTabs, {
      attachToDocument: true,
      propsData: { showArrows: true }
    })

    expect(wrapper.vm.genIcon('prepend')).toBe(null)

    // Mock display state
    wrapper.setData({ isOverflowing: true })
    wrapper.vm.$vuetify.breakpoint.width = 800
    await ssrBootable()
    wrapper.setProps({ mobileBreakPoint: 1200 })

    expect(wrapper.vm.genIcon('prepend')).toBeTruthy()
    await wrapper.vm.$nextTick()

    wrapper.setMethods({ scrollTo })

    const next = wrapper.find('.icon--append')[0]
    next.trigger('click')
    await wrapper.vm.$nextTick()
    expect(scrollTo).toHaveBeenCalledWith('append')
  })

  it('should call on touch methods', async () => {
    const wrapper = mount(VTabs, {
      attachToDocument: true
    })

    wrapper.setData({ isOverflowing: true })

    const onTouch = jest.fn()
    wrapper.setMethods({
      onTouchStart: onTouch,
      onTouchMove: onTouch,
      onTouchEnd: onTouch
    })
    await ssrBootable()

    const tabsWrapper = wrapper.find('.tabs__wrapper')[0]

    touch(tabsWrapper).start(0, 0)
    touch(tabsWrapper).end(0, 0)
    touch(tabsWrapper).move(15, 15)
    expect(onTouch.mock.calls.length).toBe(3)
  })

  it('should use a slotted slider', () => {
    const wrapper = mount(VTabs, {
      slots: {
        default: [{
          name: 'v-tabs-slider',
          render: h => h(VTabsSlider, {
            props: { color: 'pink' }
          })
        }]
      }
    })

    const slider = wrapper.find(VTabsSlider)[0]
    expect(slider.hasClass('pink')).toBe(true)
  })

  it('should call append or prepend offset method', () => {
    const wrapper = mount(VTabs)
    const offsetFn = jest.fn()

    wrapper.setMethods({
      newOffsetPrepend: offsetFn,
      newOffsetAppend: offsetFn
    })

    wrapper.vm.newOffset('append')
    wrapper.vm.newOffset('prepend')

    expect(offsetFn.mock.calls.length).toBe(2)
  })

  it('should return object or null when calling new offset', async () => {
    const wrapper = mount(VTabs)
    const mockEls = createRange(25).map(x => ({ clientWidth: 50 }))

    await ssrBootable()

    // Mocking container and children
    expect(wrapper.vm.newOffsetAppend(
      { scrollWidth: 800, clientWidth: 400 },
      mockEls,
      0
    )).toEqual({ offset: 400, index: 8 })

    // Mock offset
    wrapper.setData({ itemOffset: 10 })

    // Mocking container and children
    expect(wrapper.vm.newOffsetPrepend(
      { clientWidth: 200 },
      mockEls,
      50
    )).toEqual({ offset: -150, index: 8 })

    wrapper.setData({ itemOffset: 0 })
    // Mocking container and children
    expect(wrapper.vm.newOffsetPrepend(
      { clientWidth: 0 },
      [{ clientWidth: 50 }]
    )).toEqual({ offset: 0, index: 0 })
  })

  it('should handle touch events and remove container transition', async () => {
    const wrapper = mount(VTabs, {
      attachToDocument: true
    })

    wrapper.setData({ isOverflowing: true })
    const container = wrapper.find('.tabs__container')[0]

    await ssrBootable()

    expect(wrapper.vm.startX).toBe(0)
    wrapper.vm.onTouchStart({ touchstartX: 0 })
    expect(container.hasStyle('transition', 'none')).toBe(true)

    wrapper.vm.onTouchMove({ touchmoveX: -100 })
    expect(wrapper.vm.scrollOffset).toBe(100)

    wrapper.vm.onTouchEnd()
    expect(wrapper.vm.scrollOffset).toBe(0)

    wrapper.setData({ isOverflowing: false })
    wrapper.vm.onTouchEnd()
    expect(wrapper.vm.scrollOffset).toBe(0)
  })

  it('should generate a v-tabs-items if none present and has v-tab-item', async () => {
    const wrapper = mount(VTabs, {
      propsData: { value: 'foo' },
      slots: {
        default: [{
          name: 'v-tab-item',
          render: h => h('div')
        }]
      }
    })

    await ssrBootable()

    expect(wrapper.find(VTabsItems).length).toBe(1)
  })
})
