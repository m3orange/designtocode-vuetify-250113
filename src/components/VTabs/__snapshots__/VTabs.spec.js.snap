// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTabs should show tabs arrows 1`] = `

<div class="tabs tabs--overflow tabs--show-arrows">
  <div class="tabs__bar">
    <!---->
    <div class="tabs__wrapper">
      <div class="tabs__container"
           style="height: 48px;"
      >
        <div class="tabs__slider-wrapper">
          <div class="tabs__slider accent">
          </div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;
